### PR TITLE
[language][move] Add parse_comma_list function to the parser

### PR DESCRIPTION
**Pull request by bob-wilson**
This is the move-lang equivalent of https://github.com/libra/libra/pull/1931. In the same way that the IR compiler change made some commas mandatory, this does the same for function parameters, where all commas were previously optional.

This PR also changes to consistently allow optional trailing commas in all comma-separated lists.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

I updated the test for function parameter parsing since commas are no longer optional.

### CODE REVIEW

File: language/move-lang/src/parser/syntax.rs [hunk 1]
```diff
@@ -98,6 +98,31 @@ fn consume_optional_token_with_loc<'input>(
     }
 }
 
+fn parse_comma_list<'input, F, R>(
```
Comment (vgao1996) [inline language/move-lang/src/parser/syntax.rs]: Just a reminder that later when you run into parsing << and >>, you may want to make this even more generic. 

This is what I had to do for the IR parser:
```
fn parse_comma_list<'input, F, G, R>(
    tokens: &mut Lexer<'input>,
    is_end_of_list: G,
    parse_list_item: F,
    allow_trailing_comma: bool,
) -> Result<Vec<R>, ParseError<usize, anyhow::Error>>
where
    F: Fn(&mut Lexer<'input>) -> Result<R, ParseError<usize, anyhow::Error>>,
    G: Fn(&Lexer<'input>) -> bool,
{
    let mut v = vec![];
    if !is_end_of_list(tokens) {
        loop {
            v.push(parse_list_item(tokens)?);
            if is_end_of_list(tokens) {
                break;
            }
            consume_token(tokens, Tok::Comma)?;
            if is_end_of_list(tokens) && allow_trailing_comma {
                break;
            }
        }
    }
    Ok(v)
}
```

Comment (bob-wilson) [reply language/move-lang/src/parser/syntax.rs]: I'm planning to add an error message saying "expected '{}'" for the list end token, so using an "is_end_of_list" function won't work

File: language/move-lang/src/parser/syntax.rs [hunk 2]
```diff
@@ -537,18 +547,15 @@ fn parse_term<'input>(tokens: &mut Lexer<'input>) -> Result<Exp, ParseError> {
...
                         consume_token(tokens, Tok::Comma)?;
-                        if tokens.peek() == Tok::RParen {
-                            break;
-                        }
-                        es.push(parse_exp(tokens)?);
                     }
-                    tokens.advance()?; // consume the RParen
-                    if es.len() == 1 {
-                        es.pop().unwrap().value
+                    let mut es = parse_comma_list(tokens, Tok::RParen, parse_exp, true)?;
```
Comment (tnowacki) [inline language/move-lang/src/parser/syntax.rs]: I kind of don't like the trailing comma here, not sure what the rule was before

File: language/move-lang/src/parser/syntax.rs [hunk 3]
```diff
@@ -619,15 +619,8 @@ fn parse_pack_or_call<'input>(tokens: &mut Lexer<'input>) -> Result<Exp_, ParseE
...
-    let mut args = vec![];
-    while tokens.peek() != Tok::RParen {
-        args.push(parse_exp(tokens)?);
-        if tokens.peek() == Tok::RParen {
-            break;
-        }
-        consume_token(tokens, Tok::Comma)?;
-    }
-    tokens.advance()?; // consume the RParen
+    let args = parse_comma_list(tokens, Tok::RParen, parse_exp, true)?;
```
Comment (tnowacki) [inline language/move-lang/src/parser/syntax.rs]: I kind of don't like the trailing comma on ExpList, and I feel like I shouldn't like this one for consistency, but I'm unsure

File: language/move-lang/src/parser/syntax.rs [hunk 4]
```diff
@@ -1058,31 +1033,18 @@ fn parse_function_decl<'input>(
...
-            type_parameters.push(parse_type_parameter(tokens)?);
-            if tokens.peek() == Tok::Greater {
-                break;
-            }
-            consume_token(tokens, Tok::Comma)?;
-        }
-        tokens.advance()?; // consume the ">"
-    }
+    let type_parameters = if has_params {
+        let list = parse_comma_list(tokens, Tok::Greater, parse_type_parameter, true)?;
```
Comment (tnowacki) [inline language/move-lang/src/parser/syntax.rs]: Should be false?

File: language/move-lang/src/parser/syntax.rs [hunk 5]
```diff
@@ -1058,31 +1033,18 @@ fn parse_function_decl<'input>(
...
-        let v = parse_var(tokens)?;
-        consume_token(tokens, Tok::Colon)?;
-        let t = parse_single_type(tokens)?;
-        if tokens.peek() == Tok::Comma {
-            tokens.advance()?;
-        }
-        parameters.push((v, t));
-    }
-    tokens.advance()?; // consume the ")"
+    let parameters = parse_comma_list(tokens, Tok::RParen, parse_parameter, true)?;
```
Comment (tnowacki) [inline language/move-lang/src/parser/syntax.rs]: This one should probably be consistent with call (which it is now, just highlighting for any changes)

File: language/move-lang/src/parser/syntax.rs [hunk 6]
```diff
@@ -1163,17 +1133,13 @@ fn parse_struct_definition<'input>(
...
-            type_parameters.push(parse_type_parameter(tokens)?);
-            if tokens.peek() == Tok::Greater {
-                break;
-            }
-            consume_token(tokens, Tok::Comma)?;
-        }
-        tokens.advance()?; // consume the ">"
-    }
+    let type_parameters = if has_params {
+        let list = parse_comma_list(tokens, Tok::Greater, parse_type_parameter, true)?;
```
Comment (tnowacki) [inline language/move-lang/src/parser/syntax.rs]: Should be false?

**Comment by bob-wilson** [comment]
OK. As I commented in the initial description of this PR, I was planning to tackle any language changes separately, but if you'd prefer we can include that here. I suspect there may be some bikeshedding involved....

I did some reading about where and when to allow/use trailing commas. Here are a few highlights:

- People have strong preferences here (in both directions) so giving them the option of using trailing commas may be valuable.
- When code is generated automatically, it is somewhat easier to just emit a comma after each item in a list. I don't expect this to be very relevant for Move, and it's not a very strong argument anyway.
- When items are on separate lines, trailing commas make it easier to modify the list (i.e., without having to add a comma to the previously final entry). Most arguments about this topic focus on the multi-line situation. I don't remember seeing anyone arguing for using trailing commas when a list is on one source line.
- Some languages and/or code formatting tools look at whitespace and distinguish the case where there is a newline after a list item, requiring/using a trailing comma then but not otherwise. For the sake of simplicity, I'd rather not go that route, but it is an option.

I reviewed the different kinds of comma-separated lists in Move:

- Fields (in struct definitions, bindings, and pack operations)
- Function arguments (both definitions and calls)
- Type parameters for generics
- Things related to multiple return values: Expressions (in expression lists), Types (in tuple-like return types), and Bindings (when binding multiple return values)
- Acquires

@tnowacki suggests here that we change to ban trailing commas for everything except field lists and possibly function arguments. IMO, the function arguments should be consistent with generic type parameters, so if we're not going to allow trailing commas for type parameters, I propose that we also ban them for function arguments.

At the other extreme, we could choose to allow optional trailing commas for all of these. (It would be a minor hassle to parse that for the "acquires" lists, but not a serious problem.) That would be very consistent and leave it up to individual developers to decide which style they like. I don't necessarily want to advocate for that, but it is something we should at least consider.

Personally, I would be OK to say that trailing commas are only allowed for field lists.

**Comment by tnowacki** [comment]
>  the function arguments should be consistent with generic type parameters

I'm not sure if I follow why this is true. For me its more that function arguments need to be consistent with ExpList. I don't see it being necessary to be consistent with types. 

Definitely though, type parameters need to be consistent with type arguments

**Comment by bob-wilson** [comment]
> I'm not sure if I follow why this is true. For me its more that function arguments need to be consistent with ExpList. I don't see it being necessary to be consistent with types.

Just that when you see a generic function there are 2 comma-separated lists of parameters, one for the types and the other for the arguments. It would be inconsistent and confusing if the rules for trailing commas would be different between those 2 lists.

**Comment by tnowacki** [comment]
>  It would be inconsistent and confusing if the rules for trailing commas would be different between those 2 lists.

I don't see how its inconsistent or confusing. No more confusing than allowing trailing commas on pack but not on call

**Comment by tnowacki** [comment]
These are all different cases, I don't think it is confusing or inconsistent to have trailing commas one thing and and not on another, even if those things are next to each other in code. 

Keep in mind type arguments also exist at pack and unpack, alongside call and call. I think its totally reasonable for types to have a different rule than values/expressions

**Comment by bob-wilson** [comment]
Todd and I discussed this in person. We agreed that the decision of where to use trailing commas is best left to formatting tools, where it is more natural to consider whitespace, e.g., whether a list item is followed by a newline. Otherwise, it is best for the compiler to be consistent and permissive, so that it doesn't frustrate advanced users who have strong opinions about using trailing commas. I've gone ahead and added changes here to allow trailing commas in all comma-separated lists.

**Pull request closed without merging**